crates/aggchain-proof-service/src/service/customchaindata_builder.rs

```diff
@@ -0,0 +1,281 @@
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {}
+
+#[derive(Debug)]
+pub(crate) struct CustomChainDataBuilderRequest {
+    pub(crate) l2_block_number: u64,
+    pub(crate) output_root: [u8; 32],
+}
+
+#[derive(Debug, PartialEq)]
+pub(crate) struct CustomChainDataBuilderResponse {
+    pub custom_chain_data: [u8; 66],
+}
+
+pub(crate) struct CustomChainDataBuilderService<L1Rpc> {
+    l1_rpc: Arc<L1Rpc>,
+    network_id: u32,
+    vkey: Arc<[u32; 8]>,
+}
+
+impl Clone for CustomChainDataBuilderService<L1Rpc> {
+    fn clone(&self) -> Self {
+        CustomChainDataBuilderService {
+            l1_rpc: self.l1_rpc.clone(),
+            network_id: self.network_id,
+            vkey: self.vkey.clone(),
+        }
+    }
+}
+
+impl<L1Rpc> CustomChainDataBuilderService<L1Rpc> {
+    pub fn new(l1_rpc: Arc<L1Rpc>, network_id: u32, vkey: [u32; 8]) -> Self {
+        CustomChainDataBuilderService {
+            l1_rpc,
+            network_id,
+            vkey: Arc::new(vkey),
+        }
+    }
+}
+
+impl<L1Rpc> tower::Service<CustomChainDataBuilderRequest> for CustomChainDataBuilderService<L1Rpc>
+where
+    L1Rpc: AggchainVkeyResolver + Send + Sync + 'static,
+{
+    type Response = CustomChainDataBuilderResponse;
+    type Error = Error;
+    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, req: CustomChainDataBuilderRequest) -> Self::Future {
+        let rpc = self.l1_rpc.clone();
+
+        let vkey = self.vkey.clone();
+        async move {
+            let registered_vkey = rpc.resolve_vkey(AGGCHAIN_VKEY_SELECTOR).await?;
+            if *vkey != registered_vkey {
+                error!(
+                    "L1 registered vKey doesn't match the one configured in the \
+                     aggchain-proof-service"
+                )
+            }
+
+            Ok(CustomChainDataBuilderResponse {
+                custom_chain_data: calculate_custom_chain_data(
+                    AGGCHAIN_PROOF_PROGRAM_VERSION,
+                    req.output_root,
+                    req.l2_block_number,
+                ),
+            })
+        }
+        .boxed()
+    }
+}
+
+#[async_trait::async_trait]
+#[cfg_attr(test, mockall::automock)]
+trait AggchainVkeyResolver {
+    async fn resolve_vkey(&self, aggchain_vkey_selector: [u8; 4]) -> Result<[u32; 8], Error>;
+}
+
+pub struct L1Rpc {}
+
+#[async_trait::async_trait]
+impl AggchainVkeyResolver for L1Rpc {
+    async fn resolve_vkey(&self, _aggchain_vkey_selector: [u8; 4]) -> Result<[u32; 8], Error> {
+        // Implementation for resolving vkey from L1 RPC
+        unimplemented!()
+    }
+}
+
+pub fn calculate_custom_chain_data(
+    selector: u16,
+    output_root: [u8; 32],
+    l2_block_number: u64,
+) -> [u8; 66] {
+    let mut custom_chain_data = [0u8; 66];
```

Comment:
Is the custom chain data always in this format, i.e. selector, output root, (gap), l2 block? If so, could it just be a type with these fields that serializes / deserializes to the appropriate byte sequence?